fiona/ogrext.pyx

```diff
@@ -1285,6 +1283,15 @@ cdef class WritingSession(Session):
         self._schema_mapping = dict(zip(before_fields.keys(),
                                         after_fields.keys()))
 
+        # Mapping of the Python collection schema to OGR field indices.
+        # We assume that get_schema()['properties'].keys() is in the exact OGR field order
+        assert len(before_fields) == len(after_fields)
+        self._schema_mapping_index = dict(zip(before_fields.keys(), range(len(after_fields.keys()))))
```

Comment:
instead of looking up the field index by `OGR_F_GetFieldIndex`, using the field name and cogr_feature in `OGRFeatureBuilder.Build()`, each time we write a record, the idea is to note the order of the field definitions from when we inspect the layer schema, and store that once for reuse when writing all records.

i am not very familiar with ogr, if field indices can vary arbitrary per record, this will be defective.

Reply:
@fcostin seems like a good strategy. Field indices are per-layer.